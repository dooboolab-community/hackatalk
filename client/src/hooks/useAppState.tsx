import {AppState, AppStateStatus} from 'react-native';
import {useEffect, useState} from 'react';

export default function useAppState(): AppStateStatus {
  const currentState = AppState.currentState;
  const [appState, setAppState] = useState(currentState);

  function onChange(newState: AppStateStatus): void {
    setAppState(newState);
  }

  useEffect(() => {
    AppState.addEventListener('change', onChange);

    return (): void => {
      AppState.removeEventListener('change', onChange);
    };
  });

  return appState;
}
