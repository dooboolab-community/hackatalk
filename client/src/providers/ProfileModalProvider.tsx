import React, {FC, createContext, useContext, useState} from 'react';

import {ProfileModal_user$key} from '../__generated__/ProfileModal_user.graphql';

export type ModalState =
  | {
      isVisible: false;
    }
  | {
      /** Modal visiblity */
      isVisible: true;
      /** Which user the profile modal describes */
      user: ProfileModal_user$key;
      /** Is the profile user a friend of the current user? */
      isFriend?: boolean;
      /** Callback function for delete button */
      onDeleteFriend?: () => void;
      /** Callback function for add friend button */
      onAddFriend?: () => void;
      hideButtons?: boolean;
    };

export type ProfileModalContext = {
  /** Allow context consumers to make the modal visible with a new state. */
  showModal: (next: Omit<ModalState & {isVisible: true}, 'isVisible'>) => void;
  /** Allow context consumers to hide the modal. */
  hideModal: () => void;
  modalState: ModalState;
};

const ProfileModalContext = createContext<ProfileModalContext>({
  modalState: {isVisible: false},
  showModal: () => {},
  hideModal: () => {},
});

export const ProfileModalProvider: FC = (props) => {
  const [state, setState] = useState<ModalState>({
    isVisible: false,
  });

  const showModal = (
    next: Omit<ModalState & {isVisible: true}, 'isVisible'>,
  ): void =>
    setState({
      isVisible: true,
      ...next,
    });

  const hideModal = (): void =>
    setState({
      isVisible: false,
    });

  return (
    <ProfileModalContext.Provider
      value={{
        showModal,
        hideModal,
        modalState: state,
      }}>
      {props.children}
    </ProfileModalContext.Provider>
  );
};

export const useProfileContext = (): ProfileModalContext =>
  useContext(ProfileModalContext);
