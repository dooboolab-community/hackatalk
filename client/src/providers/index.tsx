export * from './AuthProvider';
export * from './DeviceProvider';
export * from './ProfileModalProvider';
