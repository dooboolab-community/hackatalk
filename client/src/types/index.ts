import { StyleProp, TextStyle } from 'react-native';

import { SFC } from 'react';
import { User } from './graphql';

export enum MessageType {
  Message,
  Photo,
  File,
}

export interface Friend extends User {
  isFriend?: boolean;
  friendSince: Date;
}

interface IconProps {
  style?: StyleProp<TextStyle>;
  width?: number | string;
  height?: number | string;
  children?: never;
}

export type IconType = SFC<IconProps>;

export enum ThemeType {
  LIGHT = 'LIGHT',
  DARK = 'DARK',
}
