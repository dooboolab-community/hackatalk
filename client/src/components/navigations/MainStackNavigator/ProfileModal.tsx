import {Alert, TouchableOpacity, View, ViewStyle} from 'react-native';
import {
  ChannelFindOrCreatePrivateChannelMutation,
  ChannelFindOrCreatePrivateChannelMutationResponse,
} from '../../../__generated__/ChannelFindOrCreatePrivateChannelMutation.graphql';
import {LoadingIndicator, useTheme} from 'dooboo-ui';
import {
  ProfileModalContext,
  useProfileContext,
} from '../../../providers/ProfileModalProvider';
import React, {FC, useState} from 'react';
import {
  addFriendMutation,
  deleteFriendMutation,
} from '../../../relay/queries/Friend';
import {
  createBlockedUserMutation,
  deleteBlockedUserMutation,
} from '../../../relay/queries/BlockedUser';

import {BlockedUserCreateMutation} from '../../../__generated__/BlockedUserCreateMutation.graphql';
import {BlockedUserDeleteMutation} from '../../../__generated__/BlockedUserDeleteMutation.graphql';
import {ConnectionHandler} from 'relay-runtime';
import {FontAwesome} from '@expo/vector-icons';
import {FriendAddMutation} from '../../../__generated__/FriendAddMutation.graphql';
import {FriendDeleteMutation} from '../../../__generated__/FriendDeleteMutation.graphql';
import {IC_NO_IMAGE} from '../../../utils/Icons';
import Modal from 'react-native-modalbox';
import {findOrCreatePrivateChannel} from '../../../relay/queries/Channel';
import {getString} from '../../../../STRINGS';
import {showAlertForError} from '../../../utils/common';
import styled from 'styled-components/native';
import {useMutation} from 'react-relay/hooks';
import {useNavigation} from '@react-navigation/core';

const StyledView = styled.View`
  margin-top: 64px;
`;

const StyledImage = styled.Image`
  width: 80px;
  height: 80px;
  border-radius: 40px;
`;

const StyledViewBtns = styled.View`
  height: 48px;
  align-self: stretch;
  background-color: ${({theme}) => theme.modalBtnBackground};
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
`;

const StyledViewBtnDivider = styled.View`
  width: 0.5px;
  height: 48px;
  background-color: ${({theme}) => theme.placeholder};
`;

const StyledTextDisplayName = styled.Text`
  font-size: 24px;
  color: white;
  font-weight: bold;
  margin-top: 16px;
  padding: 0 32px;
  align-self: center;
`;

const StyledTextstatusMessage = styled.Text`
  font-size: 12px;
  color: white;
  margin-top: 8px;
  align-self: center;
`;

const StyledText = styled.Text`
  color: ${({theme}) => theme.modalBtnFont};
  font-size: 16px;
`;

const StyledTextFriendAdded = styled.Text`
  color: ${({theme}) => theme.tintColor};
  font-size: 12px;
  background-color: ${({theme}) => theme.background};
  padding: 4px;
`;

const StyledTextFriendAlreadyAdded = styled.Text`
  color: ${({theme}) => theme.negative};
  font-size: 12px;
  background-color: ${({theme}) => theme.background};
  padding: 4px;
`;

interface Styles {
  wrapper: ViewStyle;
  viewBtn: ViewStyle;
}

const styles: Styles = {
  wrapper: {
    backgroundColor: 'transparent',
    alignSelf: 'stretch',
    height: 320,
    width: '90%',
    alignContent: 'center',
    alignItems: 'center',
    justifyContent: 'center',
  },
  viewBtn: {
    width: '50%',
    height: 80,
    alignItems: 'center',
    justifyContent: 'center',
  },
};

type ModalContentProps = {
  isVisible: boolean;
} & ProfileModalContext;

const ModalContent: FC<ModalContentProps> = ({
  modalState,
  hideModal,
}: ModalContentProps) => {
  const [showFriendAddedMessage, setShowFriendAddedMessage] = useState<boolean>(
    false,
  );

  const navigation = useNavigation();

  const [
    commitChannel,
    isChannelInFlight,
  ] = useMutation<ChannelFindOrCreatePrivateChannelMutation>(
    findOrCreatePrivateChannel,
  );

  const [commitAddFriend, addFriendInFlight] = useMutation<FriendAddMutation>(
    addFriendMutation,
  );

  const [
    commitDeleteFriend,
    deleteFriendInFlight,
  ] = useMutation<FriendDeleteMutation>(deleteFriendMutation);

  const [
    commitCreateBlockedUser,
    isCreateBlockedUserInFlight,
  ] = useMutation<BlockedUserCreateMutation>(createBlockedUserMutation);

  const [
    commitDeleteBlockedUser,
    isDeleteBlockedUserInFlight,
  ] = useMutation<BlockedUserDeleteMutation>(deleteBlockedUserMutation);

  const addFriend = async (): Promise<void> => {
    if (modalState) {
      const {user, onAddFriend} = modalState;

      commitAddFriend({
        variables: {friendId: user.id},
        updater: (proxyStore) => {
          const root = proxyStore.getRoot();

          const connectionRecord =
            root && ConnectionHandler.getConnection(root, 'Friend_friends');

          const userProxy = proxyStore.get(user.id);

          const newEdge =
            connectionRecord &&
            userProxy &&
            ConnectionHandler.createEdge(
              proxyStore,
              connectionRecord,
              userProxy,
              'User',
            );

          if (connectionRecord && newEdge)
            ConnectionHandler.insertEdgeAfter(connectionRecord, newEdge);
        },
      });

      if (onAddFriend) onAddFriend();

      setShowFriendAddedMessage(true);
    }
  };

  const deleteFriend = async (): Promise<void> => {
    if (modalState) {
      const {user, onDeleteFriend} = modalState;

      commitDeleteFriend({
        variables: {friendId: user.id},
        updater: (proxyStore) => {
          const root = proxyStore.getRoot();

          const connectionRecord =
            root && ConnectionHandler.getConnection(root, 'Friend_friends');

          if (connectionRecord)
            ConnectionHandler.deleteNode(connectionRecord, user.id);
        },
      });

      if (onDeleteFriend) onDeleteFriend();

      hideModal();
    }
  };

  const createBlockedUser = (): void => {
    const blockedUserId = modalState?.user?.id;

    if (blockedUserId)
      commitCreateBlockedUser({
        variables: {blockedUserId},
      });

    hideModal();
  };

  const deleteBlockedUser = (): void => {
    const blockedUserId = modalState?.user?.id;

    if (blockedUserId)
      commitDeleteBlockedUser({
        variables: {blockedUserId},
      });

    hideModal();
  };

  const startChatting = (): void => {
    const user = modalState?.user;

    if (user) {
      const mutationConfig = {
        variables: {
          peerUserIds: [user.id],
        },
        onCompleted: (
          response: ChannelFindOrCreatePrivateChannelMutationResponse,
        ): void => {
          const channel = response.findOrCreatePrivateChannel;

          hideModal();

          navigation.navigate('Message', {
            users: [user],
            channel,
          });
        },
        onError: (error: Error): void => {
          showAlertForError(error);
        },
      };

      commitChannel(mutationConfig);
    }
  };

  const {
    theme: {header, modalBtnPrimaryFont},
  } = useTheme();

  const imageURL =
    typeof modalState?.user.photoURL === 'string' &&
    modalState?.user.photoURL !== 'null'
      ? {uri: modalState?.user.photoURL}
      : modalState?.user.photoURL;

  return (
    <View
      style={{
        height: 300,
        marginHorizontal: 20,
        alignSelf: 'stretch',
        alignItems: 'center',
        justifyContent: 'space-between',
        backgroundColor: header,
      }}>
      <View
        style={{
          position: 'absolute',
          top: 4,
          right: 8,
          flexDirection: 'row',
        }}>
        <TouchableOpacity
          testID="touch-done"
          onPress={() => {
            navigation.navigate('Report', {
              name:
                modalState?.user.nickname ||
                modalState?.user.name ||
                getString('NO_NAME'),
              userId: modalState?.user.id,
            });

            hideModal();
          }}>
          <View
            style={{
              paddingRight: 12,
              paddingLeft: 8,
              paddingVertical: 8,
            }}>
            <FontAwesome name="exclamation-circle" size={24} color="white" />
          </View>
        </TouchableOpacity>
        {isCreateBlockedUserInFlight || isDeleteBlockedUserInFlight ? (
          <View
            style={{
              paddingRight: 16,
              paddingLeft: 8,
              paddingVertical: 8,
              justifyContent: 'center',
            }}>
            <LoadingIndicator size="small" />
          </View>
        ) : (
          <TouchableOpacity
            testID="touch-done"
            onPress={(): void =>
              Alert.alert(
                modalState?.user.hasBlocked
                  ? getString('UNBAN_USER')
                  : getString('BAN_USER'),
                modalState?.user.hasBlocked
                  ? getString('UNBAN_USER_TEXT')
                  : getString('BAN_USER_TEXT'),
                [
                  {
                    text: getString('NO'),
                    onPress: () => {},
                    style: 'cancel',
                  },
                  {
                    text: getString('YES'),
                    onPress: modalState?.user.hasBlocked
                      ? deleteBlockedUser
                      : createBlockedUser,
                  },
                ],
                {cancelable: false},
              )
            }>
            <View
              style={{
                paddingRight: 16,
                paddingLeft: 8,
                paddingVertical: 8,
              }}>
              <FontAwesome
                name="ban"
                size={24}
                color={modalState?.user.hasBlocked ? 'red' : 'white'}
              />
            </View>
          </TouchableOpacity>
        )}
      </View>
      <StyledView>
        <TouchableOpacity
          activeOpacity={0.5}
          onPress={() => {
            const user = modalState?.user;

            if (user)
              navigation.navigate('ImageSlider', {
                images: [{uri: user.photoURL, sender: user.name}],
              });
          }}>
          {modalState?.user.photoURL ? (
            <StyledImage
              style={{alignSelf: 'center'}}
              source={imageURL || IC_NO_IMAGE}
            />
          ) : (
            <View
              style={{
                width: 80,
                height: 80,
                alignSelf: 'center',
                alignItems: 'center',
                justifyContent: 'center',
              }}>
              <StyledImage style={{alignSelf: 'center'}} source={IC_NO_IMAGE} />
            </View>
          )}
        </TouchableOpacity>
        <StyledTextDisplayName numberOfLines={1}>
          {modalState?.user.nickname ||
            modalState?.user.name ||
            getString('NO_NAME')}
        </StyledTextDisplayName>
        <StyledTextstatusMessage>
          {modalState?.user.statusMessage}
        </StyledTextstatusMessage>
      </StyledView>
      {showFriendAddedMessage ? (
        addFriendInFlight ? (
          <LoadingIndicator size="small" />
        ) : modalState?.isFriend ? (
          <StyledTextFriendAlreadyAdded testID="already-added-message">
            {getString('FRIEND_ALREADY_ADDED')}
          </StyledTextFriendAlreadyAdded>
        ) : (
          <StyledTextFriendAdded testID="added-message">
            {getString('FRIEND_ADDED')}
          </StyledTextFriendAdded>
        )
      ) : null}
      {!modalState?.hideButtons ? (
        <StyledViewBtns>
          {deleteFriendInFlight ? (
            <LoadingIndicator size="small" />
          ) : (
            <TouchableOpacity
              testID="touch-add-friend"
              activeOpacity={0.5}
              onPress={modalState?.isFriend ? deleteFriend : addFriend}
              style={styles.viewBtn}>
              <View style={styles.viewBtn}>
                <StyledText testID="text-add-title">
                  {modalState?.isFriend
                    ? getString('DELETE_FRIEND')
                    : getString('ADD_FRIEND')}
                </StyledText>
              </View>
            </TouchableOpacity>
          )}
          <StyledViewBtnDivider />
          <TouchableOpacity
            testID="btn-chat"
            activeOpacity={0.5}
            onPress={startChatting}
            style={styles.viewBtn}>
            {isChannelInFlight ? (
              <LoadingIndicator size="small" />
            ) : (
              <View style={styles.viewBtn}>
                <StyledText
                  style={{
                    color: modalBtnPrimaryFont,
                  }}>
                  {getString('CHAT')}
                </StyledText>
              </View>
            )}
          </TouchableOpacity>
        </StyledViewBtns>
      ) : null}
    </View>
  );
};

interface Props {
  testID?: string;
}

const ProfileModal: FC<Props> = () => {
  const profileContext = useProfileContext();
  const {isVisible, hideModal} = profileContext;

  return (
    <Modal
      isOpen={isVisible}
      backdropOpacity={0.075}
      entry={'top'}
      position={'center'}
      /*
       * `hideModal` should be called on closed event,
       * because `Modal` cannot update its props by itself.
       * If `hideModal` is not called on closed event,
       * `Modal` gets into an illegal state where `isOpen`
       * props is true while the internal state is closed.
       */
      onClosed={hideModal}
      style={styles.wrapper}>
      {profileContext.isVisible ? <ModalContent {...profileContext} /> : null}
    </Modal>
  );
};

export default ProfileModal;
