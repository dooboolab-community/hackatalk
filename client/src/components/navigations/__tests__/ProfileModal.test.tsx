import {MockPayloadGenerator, createMockEnvironment} from 'relay-test-utils';
import React, {
  FC,
  RefObject,
  createRef,
  forwardRef,
  useImperativeHandle,
} from 'react';
import {act, fireEvent, render, waitFor} from '@testing-library/react-native';
import {
  createMockNavigation,
  createTestElement,
} from '../../../../test/testUtils';
import {graphql, useLazyLoadQuery} from 'react-relay/hooks';

import {IEnvironment} from 'relay-runtime';
import ProfileModal from '../MainStackNavigator/ProfileModal';
import {ProfileModalTestQuery} from '../../../__generated__/ProfileModalTestQuery.graphql';
import ReactNavigation from '@react-navigation/core';
import {User} from '../../../types/graphql';
import {View} from 'react-native';
import {useProfileContext} from '../../../providers/ProfileModalProvider';

const mockNavigation = createMockNavigation();

jest.mock('@react-navigation/core', () => ({
  ...jest.requireActual<typeof ReactNavigation>('@react-navigation/core'),
  useNavigation: () => mockNavigation,
}));

const generateUser = (idNum: number, isFriend: boolean): Partial<User> => ({
  id: `user-test-${idNum}`,
  isFriend,
});

type ConsumerRef = {
  showModal: (next: {
    onDeleteFriend?: () => void;
    onAddFriend?: () => void;
    hideButtons?: boolean;
  }) => void;
  hideModal: () => void;
};

const ProfileConsumer = forwardRef<ConsumerRef>((_props, ref) => {
  const {myData} = useLazyLoadQuery<ProfileModalTestQuery>(
    graphql`
      query ProfileModalTestQuery {
        myData: user(id: "test-id") {
          ...ProfileModal_user
        }
      }
    `,
    {},
  );

  const {showModal, hideModal} = useProfileContext();

  if (!myData) throw new Error('myData is null');

  useImperativeHandle(ref, () => ({
    showModal: (next) => showModal({...next, user: myData}),
    hideModal,
  }));

  return null;
});

interface TestComponentProps {
  consumerRef: RefObject<ConsumerRef>;
  environment: IEnvironment;
}

const TestComponent: FC<TestComponentProps> = ({consumerRef, environment}) => {
  return createTestElement(
    <View>
      <ProfileConsumer ref={consumerRef} />
      <ProfileModal />
    </View>,
    {
      environment,
    },
  );
};

describe('[ProfileModal] rendering test', () => {
  it('Render without crashing', async () => {
    const mockEnvironment = createMockEnvironment();

    mockEnvironment.mock.queueOperationResolver((operation) =>
      MockPayloadGenerator.generate(operation, {
        User: (_, generateId) => generateUser(generateId(), false),
      }),
    );

    const consumerRef = createRef<ConsumerRef>();

    const screen = render(
      <TestComponent consumerRef={consumerRef} environment={mockEnvironment} />,
    );

    act(() => consumerRef.current?.showModal({}));

    const json = screen.toJSON();

    expect(json).toMatchSnapshot();
  });

  it('Should be opened', async () => {
    const mockEnvironment = createMockEnvironment();

    mockEnvironment.mock.queueOperationResolver((operation) =>
      MockPayloadGenerator.generate(operation, {
        User: (_, generateId) => generateUser(generateId(), false),
      }),
    );

    const consumerRef = createRef<ConsumerRef>();

    const screen = render(
      <TestComponent consumerRef={consumerRef} environment={mockEnvironment} />,
    );

    act(() => consumerRef.current?.showModal({}));

    const button = screen.getByTestId('touch-add-friend');

    expect(button).toBeTruthy();
  });

  it('Check "Added to your friend." button', async () => {
    const mockEnvironment = createMockEnvironment();

    mockEnvironment.mock.queueOperationResolver((operation) =>
      MockPayloadGenerator.generate(operation, {
        User: (_, generateId) => generateUser(generateId(), false),
      }),
    );

    const consumerRef = createRef<ConsumerRef>();

    const screen = render(
      <TestComponent consumerRef={consumerRef} environment={mockEnvironment} />,
    );

    act(() => consumerRef.current?.showModal({}));

    const button = screen.getByTestId('touch-add-friend');

    fireEvent.press(button);

    // Resolve mutation.
    act(() => {
      mockEnvironment.mock.resolveMostRecentOperation((operation) =>
        MockPayloadGenerator.generate(operation, {
          User: (_, generateId) => generateUser(generateId(), true),
        }),
      );
    });

    const message = screen.getByTestId('added-message');

    expect(message).toBeTruthy();
  });

  it('Should be closed', async () => {
    const mockEnvironment = createMockEnvironment();

    mockEnvironment.mock.queueOperationResolver((operation) =>
      MockPayloadGenerator.generate(operation, {
        User: (_, generateId) => generateUser(generateId(), false),
      }),
    );

    const consumerRef = createRef<ConsumerRef>();

    const screen = render(
      <TestComponent consumerRef={consumerRef} environment={mockEnvironment} />,
    );

    act(() => consumerRef.current?.hideModal());

    const button = screen.queryByTestId('touch-add-friend');

    expect(button).toBeNull();
  });

  it('delete', async () => {
    const mockEnvironment = createMockEnvironment();

    mockEnvironment.mock.queueOperationResolver((operation) =>
      MockPayloadGenerator.generate(operation, {
        User: (_, generateId) => generateUser(generateId(), true),
      }),
    );

    const consumerRef = createRef<ConsumerRef>();

    const screen = render(
      <TestComponent consumerRef={consumerRef} environment={mockEnvironment} />,
    );

    const mockOnDeleteFriend = jest.fn();

    act(() =>
      consumerRef.current?.showModal({
        onDeleteFriend: mockOnDeleteFriend,
      }),
    );

    // Find the delete button.
    // 'touch-add-friend' button becomes the delete button if
    // the user is already a friend.
    const button = screen.getByTestId('touch-add-friend');

    fireEvent.press(button);
    expect(mockOnDeleteFriend).toHaveBeenCalledTimes(1);
  });

  it('Add Friend', async () => {
    const mockEnvironment = createMockEnvironment();

    mockEnvironment.mock.queueOperationResolver((operation) =>
      MockPayloadGenerator.generate(operation, {
        User: (_, generateId) => generateUser(generateId(), false),
      }),
    );

    const consumerRef = createRef<ConsumerRef>();

    const screen = render(
      <TestComponent consumerRef={consumerRef} environment={mockEnvironment} />,
    );

    const mockOnAddFriend = jest.fn();

    act(() => {
      consumerRef.current?.showModal({
        onAddFriend: mockOnAddFriend,
      });
    });

    const button = screen.getByTestId('touch-add-friend');

    fireEvent.press(button);

    expect(mockOnAddFriend).toHaveBeenCalledTimes(1);
  });
});
