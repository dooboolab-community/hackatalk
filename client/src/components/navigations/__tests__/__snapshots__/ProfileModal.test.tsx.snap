// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[ProfileModal] rendering test Render without crashing 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View>
    <View
      accessibilityViewIsModal={true}
      importantForAccessibility="yes"
      pointerEvents="box-none"
      style={
        Array [
          Object {
            "backgroundColor": "rgba(0,0,0,0)",
            "zIndex": 2,
          },
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
        ]
      }
    >
      <View
        onLayout={[Function]}
        pointerEvents="box-none"
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          importantForAccessibility="no"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "opacity": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
                Object {
                  "backgroundColor": "black",
                  "opacity": 0.075,
                },
              ]
            }
          />
        </View>
        <View
          collapsable={false}
          onLayout={[Function]}
          onMoveShouldSetResponder={[Function]}
          onMoveShouldSetResponderCapture={[Function]}
          onResponderEnd={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderReject={[Function]}
          onResponderRelease={[Function]}
          onResponderStart={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          onStartShouldSetResponderCapture={[Function]}
          style={
            Object {
              "alignContent": "center",
              "alignItems": "center",
              "alignSelf": "stretch",
              "backgroundColor": "white",
              "height": "40%",
              "justifyContent": "center",
              "maxHeight": 320,
              "maxWidth": 480,
              "transform": Array [
                Object {
                  "translateY": -1334,
                },
                Object {
                  "translateX": 0,
                },
              ],
              "width": "90%",
            }
          }
        >
          <View
            onLayout={[Function]}
            style={
              Object {
                "alignItems": "center",
                "alignSelf": "stretch",
                "backgroundColor": "#000",
                "flex": 1,
                "justifyContent": "space-between",
              }
            }
            testID="container"
          >
            <View
              style={
                Object {
                  "flexDirection": "row",
                  "position": "absolute",
                  "right": 8,
                  "top": 4,
                }
              }
            >
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
                testID="touch-done"
              >
                <View
                  style={
                    Object {
                      "paddingLeft": 8,
                      "paddingRight": 12,
                      "paddingVertical": 8,
                    }
                  }
                >
                  <Text />
                </View>
              </View>
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
                testID="touch-done"
              >
                <View
                  style={
                    Object {
                      "paddingLeft": 8,
                      "paddingRight": 16,
                      "paddingVertical": 8,
                    }
                  }
                >
                  <Text />
                </View>
              </View>
            </View>
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "marginTop": 64,
                    },
                    undefined,
                  ]
                }
              >
                <Image
                  source={
                    Object {
                      "uri": "<mock-value-for-field-\\"photoURL\\">",
                    }
                  }
                  style={
                    Array [
                      Object {
                        "borderBottomLeftRadius": 40,
                        "borderBottomRightRadius": 40,
                        "borderTopLeftRadius": 40,
                        "borderTopRightRadius": 40,
                        "height": 80,
                        "width": 80,
                      },
                      Object {
                        "alignSelf": "center",
                      },
                    ]
                  }
                />
                <Text
                  numberOfLines={1}
                  style={
                    Array [
                      Object {
                        "alignSelf": "center",
                        "color": "white",
                        "fontSize": 24,
                        "fontWeight": "bold",
                        "marginTop": 16,
                        "paddingBottom": 0,
                        "paddingLeft": 32,
                        "paddingRight": 32,
                        "paddingTop": 0,
                      },
                      undefined,
                    ]
                  }
                >
                  &lt;mock-value-for-field-"nickname"&gt;
                </Text>
              </View>
            </View>
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "alignSelf": "stretch",
                    "backgroundColor": "#151C22",
                    "flexDirection": "row",
                    "height": 48,
                    "justifyContent": "space-between",
                  },
                  undefined,
                ]
              }
            >
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "height": "100%",
                    "justifyContent": "center",
                    "opacity": 1,
                    "width": "50%",
                  }
                }
                testID="touch-add-friend"
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "white",
                        "fontSize": 16,
                      },
                      undefined,
                    ]
                  }
                  testID="text-add-title"
                >
                  Add
                </Text>
              </View>
              <View
                style={
                  Array [
                    Object {
                      "backgroundColor": "#6D6D6D",
                      "height": 48,
                      "width": 0.5,
                    },
                    undefined,
                  ]
                }
              />
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "height": "100%",
                    "justifyContent": "center",
                    "opacity": 1,
                    "width": "50%",
                  }
                }
                testID="btn-chat"
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "white",
                        "fontSize": 16,
                      },
                      Object {
                        "color": "#fff",
                      },
                    ]
                  }
                >
                  Chat
                </Text>
              </View>
            </View>
          </View>
          <View
            collapsable={false}
            pointerEvents="box-none"
            style={
              Object {
                "backgroundColor": "black",
                "height": -48,
                "opacity": 0,
                "position": "absolute",
                "top": 0,
                "width": 0,
                "zIndex": 1,
              }
            }
            testID="view-statusMessageView"
          />
          <View
            collapsable={false}
            onLayout={[Function]}
            style={
              Object {
                "backgroundColor": "transparent",
                "overflow": "hidden",
                "position": "absolute",
                "top": 0,
                "transform": Array [
                  Object {
                    "translateY": 140,
                  },
                ],
                "width": 200,
                "zIndex": 100,
              }
            }
            testID="view-animated"
          >
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "column",
                  "justifyContent": "center",
                  "opacity": 1,
                }
              }
              testID="touchable-statusMessageView"
            >
              <Text
                ellipsizeMode="tail"
                isStatusMessageExpanded={false}
                numberOfLines={2}
                onTextLayout={[Function]}
                style={
                  Array [
                    Object {
                      "alignSelf": "center",
                      "color": "white",
                      "fontSize": 12,
                      "fontWeight": "normal",
                      "textAlign": "center",
                    },
                    undefined,
                  ]
                }
                testID="text-main"
              >
                &lt;mock-value-for-field-"statusMessage"&gt;
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
